airflow/jobs/base_job.py

```diff
@@ -110,6 +111,7 @@ def __init__(self, executor=None, heartrate=None, *args, **kwargs):
             self.heartrate = heartrate
         self.unixname = getuser()
         self.max_tis_per_query: int = conf.getint('scheduler', 'max_tis_per_query')
+        get_listener_manager().hook.on_starting()
```

Comment:
I think this is my only question: should we combine all the hookspecs in to a single hook, or would this be "clearer"/more maintainable in future as

```suggestion
        get_listener_manager().lifecycle_hook.on_starting()
```

Possible reason for wanting this: it allows us to have conflicting names for different scopes. I suspect this could be "we'll do it when we need it" since it is "internal"

Reply:
The question here is really if we need this - having conflicting names for different scopes would be more painful for the plugin author. 

I can imagine some pain points: 
- need to remember to mark method with particular hookspec
- can't implement the same hookspec in one file - especially with same-name methods, so you're forced to have multi-file plugin.
- Airflow plugin loading API gets more confusing, since you need to know a priori which hookspec the plugin implements

Seems like the easier solution would be to "namespace" them just by name. Like; `airflow_scheduler_on_starting` and `airflow_backfill_on_starting` if those were to be two different methods.

Reply:
> can't implement the same hookspec in one file - especially with same-name methods

Plugins can be classes, not just modules.

Reply:
Yes, but still, you can't easily share state - your `lifecycle` plugin class is not the same as `dag_run` plugin class